texar/core/attention_mechanism_utils.py

```diff
@@ -0,0 +1,225 @@
+    'safe_cumprod',
+    'sparsemax',
+]
+
+
+def hardmax(logits: torch.Tensor) -> torch.Tensor:
+    r"""Returns batched one-hot vectors. The depth index containing
+    the `1` is that of the maximum logit value.
+
+    Args:
+        logits: A batch tensor of logit values.
+
+    Returns:
+        A batched one-hot tensor.
+    """
+    if not isinstance(logits, torch.Tensor):
+        logits = torch.tensor(logits)
+    depth = logits.shape[-1]
+    one_hot = torch.eye(depth, dtype=torch.int64)
+    return F.embedding(torch.argmax(logits, -1), one_hot)
+
+
+def maybe_mask_score(score: torch.Tensor,
+                     score_mask_value: torch.Tensor,
+                     memory_sequence_length: Optional[torch.LongTensor]) \
+        -> torch.Tensor:
+    r"""Mask the attention score based on the masks."""
+    if memory_sequence_length is None:
+        return score
+
+    for memory_sequence_length_value in memory_sequence_length:
+        if memory_sequence_length_value <= 0:
+            raise ValueError(
+                "All values in memory_sequence_length must be greater "
+                "than zero.")
+
+    score_mask = sequence_mask(memory_sequence_length,
+                               max_len=score.shape[1])
+    score_mask_values = score_mask_value * torch.ones_like(score)
+    return torch.where(score_mask, score, score_mask_values)
+
+
+def prepare_memory(memory: torch.Tensor,
+                   memory_sequence_length: Optional[torch.LongTensor]) \
+        -> torch.Tensor:
+    r"""Convert to tensor and possibly mask ``memory``.
+
+    Args:
+        memory: tensor, shaped ``[batch_size, max_time, ...]``.
+        memory_sequence_length: integer tensor, shaped ``[batch_size]``.
+
+    Returns:
+        A (possibly masked), new ``memory``.
+
+    Raises:
+        ValueError: if ``memory`` and ``memory_sequence_length`` do not have
+        the same ``batch_size``.
+    """
+    if (memory_sequence_length is not None and
+            not isinstance(memory_sequence_length, torch.Tensor)):
+        memory_sequence_length = torch.tensor(memory_sequence_length,
+                                              dtype=torch.long)
+
+    if memory_sequence_length is None:
+        seq_len_mask = None
+    else:
+        seq_len_mask = sequence_mask(memory_sequence_length,
+                                     max_len=memory.shape[1],
+                                     dtype=memory.dtype)
+        seq_len_batch_size = memory_sequence_length.shape[0]
+
+    # Mask the memory based on the memory mask.
+    rank = memory.dim()
+    extra_ones = [1] * (rank - 2)
+    m_batch_size = memory.shape[0]
+
+    if seq_len_mask is not None:
+        if seq_len_batch_size != m_batch_size:
+            raise ValueError("memory_sequence_length and memory tensor "
+                             "batch sizes do not match.")
+        seq_len_mask = torch.reshape(
+            seq_len_mask,
+            tuple(list(seq_len_mask.shape) + extra_ones))
```

Comment:
This cast could also be replaced by tuple construction:
```python
seq_len_mask.size() + (1,) * (rank - 2)
```
And maybe `torch.reshape` could be replaced by `seq_len_mask.view`?